<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>gi_profile.pl - To find out the distribution of GI's in a profile.</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:helpdesk@jcvi.org" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#arguments">ARGUMENTS</a></li>
	<li><a href="#copyright">COPYRIGHT</a></li>
	<li><a href="#authors">AUTHORS</a></li>
</ul>

<hr name="index" />
</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>gi_profile.pl - To find out the distribution of GI's in a profile.</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>gi_profile.pl -p &lt;profile_file&gt; -g &lt;gi_file&gt;</p>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>The program takes to files from the command line: proifle file and a gi list file. The profile file should contain the NCBI taxon id (for species) in the first column. The second column shoud contains a either 1 or 0. The gi file should list a set of GI on each line of the file. The software checks if the each GI belongs to a species that is marked as 1. If so it generates an output for each GI printing 1 if the GI is present and 0 if it is absent. It ignore all the other GIs.</p>
<p>
</p>
<hr />
<h1><a name="arguments">ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="b_p_profile_file" class="item">B &lt;-p profile_file&gt;</a></strong></dt>

<dd>
<p>Format:</p>
<p>&lt;taxon id for species&gt; &lt;\t&gt; &lt;1|0&gt;</p>
</dd>
<dt><strong><a name="b_g_gi_file" class="item">B &lt;-g gi_file&gt;</a></strong></dt>

<dd>
<p>A list of GI each present in its own line.</p>
</dd>
<dt><strong><a name="b_fasta_f" class="item">B &lt;--fasta | -f&gt;</a></strong></dt>

<dd>
<p>The script can optionally create two files 'yes.faa' containing sequences with that turns out to be 1 in the result and 'no.faa' for sequences that turns out to be 0 in the result.</p>
</dd>
<dt><strong><a name="b_database_d_database_file" class="item">B &lt;--database | -d database_file&gt;</a></strong></dt>

<dd>
<p>By default the scrip will use /usr/local/archive/projects/PPP/split_genomes/blastdb/all.peptides.fa to extract sequence for the fasta file. But you can override this default using this option. Note, this is a pointer to an indexed blast database.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="copyright">COPYRIGHT</a></h1>
<p>Copyright (c) 2009 Malay K Basu &lt;<a href="mailto:malay@bioinformatics.org">malay@bioinformatics.org</a>&gt;</p>
<p>
</p>
<hr />
<h1><a name="authors">AUTHORS</a></h1>
<p>Malay K Basu &lt;<a href="mailto:malay@bioinformatics.org">malay@bioinformatics.org</a>&gt;</p>

</body>

</html>
