<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>ppp2d.pl - This is a parallel implementation of 2D Partial Phylogenetic Profile algorithm.</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:helpdesk@jcvi.org" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#mandetory_options">MANDETORY OPTIONS</a></li>
	<li><a href="#other_options">OTHER OPTIONS</a></li>
	<li><a href="#copyright">COPYRIGHT</a></li>
	<li><a href="#authors">AUTHORS</a></li>
</ul>

<hr name="index" />
</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>ppp2d.pl - This is a parallel implementation of 2D Partial Phylogenetic Profile algorithm.</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>ppp2d.pl [options] -d db_dir -t taxon -g gi &gt; result</p>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>The software run 2 dimensional phylogenetic profile starting from a single sequence. It creates multiple input profiles for this sequence by going through the BLAST result, then searches a given genome using these profiles.</p>
<p>
</p>
<hr />
<h1><a name="mandetory_options">MANDETORY OPTIONS</a></h1>
<dl>
<dt><strong><a name="d_db_path" class="item"><strong>-d | --db path</strong></a></strong></dt>

<dd>
<p>The data directory created as mentioned above.</p>
</dd>
<dt><strong><a name="g_gi" class="item"><strong>-g | --gi</strong></a></strong></dt>

<dd>
<p>A gi number for 2D phylogenetic profile.</p>
</dd>
<dt><strong><a name="t_taxon_ncbi_taxon_id" class="item"><strong>-t | --taxon ncbi_taxon_id</strong></a></strong></dt>

<dd>
<p>NCBI taxon id for the genome that will be searched for the profile. If not given the genome to which the gi belongs will be searched.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="other_options">OTHER OPTIONS</a></h1>
<dl>
<dt><strong><a name="serial" class="item"><strong>--serial</strong></a></strong></dt>

<dd>
<p>Runs the software serially.</p>
</dd>
<dt><strong><a name="threads_num" class="item"><strong>--threads num</strong></a></strong></dt>

<dd>
<p>You can run the software in SMP mode, if you have multicore processor. If you have a quad-core machine put 4 as the parameter.</p>
</dd>
<dt><strong><a name="w_workspace_path" class="item"><strong>-w |--workspace path</strong></a></strong></dt>

<dd>
<p>The name of the cluster accessible dirctory that will act as temporary space.</p>
</dd>
<dt><strong><a name="l_level_taxonomic_level" class="item"><strong>-l | --level taxonomic_level</strong></a></strong></dt>

<dd>
<p>The taxonomic level that should be searched; family, genus, species, etc. Any taxonomic rank as understood by NCBI taxonomy database can be used. The default is to do the search as taxon id level.</p>
</dd>
<dt><strong><a name="prob_probability" class="item"><strong>--prob probability</strong></a></strong></dt>

<dd>
<p>This is the probabliity for the PPP algorithm. The default is to calculate the probablity from the the given profile.</p>
</dd>
<dt><strong><a name="s_start_depth_number" class="item"><strong>-s | --start-depth number</strong></a></strong></dt>

<dd>
<p>Start the iteration after this many taxon have reached. There is no point searching the binomial probability with very few taxonomic group. Default 5.</p>
</dd>
<dt><strong><a name="e_end_depth_number" class="item"><strong>-e | --end-depth number</strong></a></strong></dt>

<dd>
<p>Stop the iteration after this many taxonomic group has been found. By default the algorithm ends at the end of the blast result file.</p>
</dd>
<dt><strong><a name="j_skip_number" class="item"><strong>-j | --skip number</strong></a></strong></dt>

<dd>
<p>This parameter allows the iteration to skip lines of the source blast result. By default the algorithm searches every blast hit. Set it to a number highter than one and your searches will be faster but may be inaccurate.</p>
</dd>
<dt><strong><a name="h_help_print_this_help_page" class="item"><strong>-h | --help</strong>
Print this help page.</a></strong></dt>

<dt><strong><a name="keep_k" class="item"><strong>--keep | -k </strong></a></strong></dt>

<dd>
<p>If given, keep the temporary files under directory 'tmp~'.</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="copyright">COPYRIGHT</a></h1>
<p>Copyright (c) 2009 Malay K Basu</p>
<p>
</p>
<hr />
<h1><a name="authors">AUTHORS</a></h1>
<p>Malay K Basu &lt;<a href="mailto:mbasu@jcvi.org">mbasu@jcvi.org</a>&gt;</p>

</body>

</html>
